### PR TITLE
Dependency injection in flutter

### LINKED ISSUE #3146: Dependency injection in Flutter (Provider package)
**Proposed title of article:**
Dependency injection in Flutter (Provider package)

**Introduction paragraph (2-3 paragraphs):**
Dependency injection in Flutter is an object-oriented technique that sends the dependencies of another object to an object.
The provider package is an easy-to-use wrapper around Inherited Widgets. It makes it easier to manage ,use and also provides a state management technique that is used for managing a piece of data around the app.

In this article we shall shall know how to implement dependency in flutter and the reasons for dependency injections .

**Key takeaways:**
-What is dependency injection ?
-Why do we need dependency injection in Flutter?
-How do you implement dependency in Flutter?
-How to get rid of flutter package.

**References:**
N/A

### PR DESCRIPTION
**Pull request by NdoloBonface**
Dependency injection in flutter(Provider Package)

### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes #3146

### CODE REVIEW

**Comment by hectorkambow** [comment]
@NdoloBonface 

[Dependency injection in flutter #3359.pdf](https://github.com/section-engineering-education/engineering-education/files/6995956/Dependency.injection.in.flutter.3359.pdf)


👋 Good afternoon - and welcome to the EngEd program.
I'll be running a final review on your PR when ready.
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.

**Comment by NdoloBonface** [comment]
@hectorkambow 
I have revisited the article and did the changes.
Thank you.

**Comment by jerimkaura** [comment]
@NdoloBonface, kindly user an image of size `600*400` and for your hero image and rename it using lowercase

**Comment by jerimkaura** [comment]
Hello @NdoloBonface, the hero image for the article is not in the correct size, make it 600*400 and I said earlier.

**Review by jerimkaura — CHANGES_REQUESTED** [review]
Make the hero image of the right size.

**Comment by NdoloBonface** [comment]
@jerimkaura ,
Done. thank you

**Comment by jerimkaura** [comment]
The image you used doesn't look free besides having an inappropriate resolution. Please try to keep images as close to 600X400 pixels (images can be resized using Pixlr Image Editor) and make sure you have the appropriate copyright to use it. Sites like Unsplash have Creative Commons images which you can use for free.

**Comment by NdoloBonface** [comment]
I will look for one and change in a minute, @jerimkaura

**Comment by NdoloBonface** [comment]
@jerimkaura , done. Thank you

File: content/articles/Dependency-injection-in-flutter/index.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+In this article, we will learn how to implement dependency injection, get rid of flutter packages, and the reasons why we need dependency injection in a flutter.
+
+### Table of content
+- [Dependency injection in Flutter](#dependency-injection-in-Flutter)
+- [Functionalities of dependency injection](#functionalities-of-dependency-injection)
+- [Reason why we need dependency injection](#Reason-why-we-need-dependency-injection)
+- [How to get rid of flutter package](#How-to-get-rid-of-flutter-package)
+- 
+### Dependency injection in Flutter
+Although dependency injection is a simple example, libraries are frequently used to abstract it from the designer. Many of these libraries take advantage of reflection (mirrors in Dart). There are issues in a flutter:
```
Comment (jerimkaura) [inline content/articles/Dependency-injection-in-flutter/index.md:15]: This statement appears incomplete "There are issues in a flutter what?"

File: content/articles/Dependency-injection-in-flutter/index.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+```
+import ‘package:inject/inject.dart’;
+@module
+class ExampleExa {
+@provide
+NewExa newExa() => OldExa();
+}
+```
+
+**Providers**
```
Comment (jerimkaura) [inline content/articles/Dependency-injection-in-flutter/index.md:66]: Kindly add more content here to make the functionalities of DI  more understandable. I feel like it's just code with no clear explanation. Besides, a beginner would not know what you are trying to achieve.

**Review by jerimkaura — CHANGES_REQUESTED** [review]
Take a look at the suggestions.

**Comment by NdoloBonface** [comment]
Done, @jerimkaura

**Review by jerimkaura — APPROVED** [review]

**Comment by ahmadmardeni1** [comment]
Please change your hero image and I will publish your article. @NdoloBonface 
You can use [this template](https://github.com/ahmadmardeni1/Canva), design your own, or use [Unsplash](https://unsplash.com/) to find one.

**Comment by NdoloBonface** [comment]
> Please change your hero image and I will publish your article. @NdoloBonface
> You can use [this template](https://github.com/ahmadmardeni1/Canva), design your own, or use [Unsplash](https://unsplash.com/) to find one.

I have decided to use the template.
Thank you

**Comment by ahmadmardeni1** [comment]
If you want to use the template, then you have to edit the text. Follow the instructions in the GitHub repo please. @NdoloBonface

**Comment by NdoloBonface** [comment]
@ahmadmardeni1, i have used a new one from Unsplash. Thanks

**Comment by ahmadmardeni1** [comment]
Welcome to Section EngEd and congrats on your first article 🎉 @NdoloBonface 

Thank you for the review @jerimkaura 
Please go over the changes I made here 88d22ce and here 88dab48

**Pull request merged**